### DISCUSSION TITLE
Static Site Generation -> Middleware -> Device type

### DISCUSSION THREAD

**Discussion by eposha** [vercel/next.js#48142] [Ideas]
### Describe the feature you'd like to request

Add possibility to get HTML markup for static pages depending on device definition with middleware. 

### Describe the solution you'd like

Since middleware is called before returning HTML, maybe it would be possible to define device and return desired HTML depending on device type

It may be necessary to generate two HTML and return the necessary one depending on the device type. 

Or add another static page generation method and call it ISR-hybrid

This would improve performance

No more rendering unnecessary components and hiding them with CSS for mobile and desktop apps

### Describe alternatives you've considered

- Hide components with CSS
- Abandon SSG in favor of SSR and device type definition in middleware

**Comment by cipriancaba** [comment]
You can easily do this with the middleware.

We're doing something similar where we have the top level `[variant]` that is a encoded string containing a/b testing info, device info, language, etc

Then, in middleware we simply rewrite the response with

`const response = NextResponse.rewrite(request.nextUrl)`

this means that an url that looks like `/about-us` will become either `/mobile/about-us` or `/desktop/about-us`. All you need to do is to generateStaticParams in your paths including the `variant` info

**Comment by eposha** [comment]
This is the approach I'm using now, and now I need to call the revalidation of the static page for each device type every time.

**Comment by patrick-mcdougle** [comment]
so basically you prefix all of your paths using a [variant] folder that you prepend in the middleware (json stringifing data or something)? Then you can use it in params / props / paths? Is that right?

**Comment by cipriancaba** [comment]
Basically yes. You can access that via the params property and you just
decode it there

**Comment by patrick-mcdougle** [comment]
@cipriancaba sorry to bother you again, but we're running into an issue where the client side ISR prefetch / hydrate is returning full html pages instead of the props we'd expect and it's causing fatal issues. Did you have to manipulate the logic a little more than just prefixing all inbound routes in your middleware? Would you be willing to share part of that middleware file here or in an email with me?
